File: Sources/XMTP/Messages/PrivateKeyBundle.swift
//
//  PrivateKeyBundle.swift
//
//
//  Created by Pat Nakajima on 11/17/22.
//

import Foundation
import XMTPProto

public typealias PrivateKeyBundle = Xmtp_MessageContents_PrivateKeyBundle

enum PrivateKeyBundleError: Error {
	case noPreKeyFound
}

extension PrivateKeyBundle {
	init(v1: PrivateKeyBundleV1) {
		self.init()
		self.v1 = v1
	}

	func encrypted(with key: SigningKey) async throws -> EncryptedPrivateKeyBundle {
		let bundleBytes = try serializedData()
		let walletPreKey = try Crypto.secureRandomBytes(count: 32)

		let signature = try await key.sign(message: Signature.enableIdentityText(key: walletPreKey))
		let cipherText = try Crypto.encrypt(signature.rawDataWithNormalizedRecovery, bundleBytes)

		var encryptedBundle = EncryptedPrivateKeyBundle()
		encryptedBundle.v1.walletPreKey = walletPreKey
		encryptedBundle.v1.ciphertext = cipherText

		return encryptedBundle
	}
}

File: library/src/main/java/org/xmtp/android/library/messages/PrivateKeyBundle.kt
package org.xmtp.android.library.messages

import com.google.protobuf.kotlin.toByteString
import kotlinx.coroutines.runBlocking
import org.xmtp.android.library.Crypto
import org.xmtp.android.library.PreEventCallback
import org.xmtp.android.library.SigningKey
import org.xmtp.android.library.XMTPException
import org.xmtp.proto.message.contents.PrivateKeyOuterClass
import java.security.SecureRandom

typealias PrivateKeyBundle = PrivateKeyOuterClass.PrivateKeyBundle

class PrivateKeyBundleBuilder {
    companion object {
        fun buildFromV1Key(v1: PrivateKeyBundleV1): PrivateKeyBundle {
            return PrivateKeyBundle.newBuilder().also {
                it.v1 = v1
            }.build()
        }
    }
}

fun PrivateKeyBundle.encrypted(
    key: SigningKey,
    preEnableIdentityCallback: PreEventCallback? = null,
): EncryptedPrivateKeyBundle {
    val bundleBytes = toByteArray()
    val walletPreKey = SecureRandom().generateSeed(32)

    preEnableIdentityCallback?.let {
        runBlocking {
            it.invoke()
        }
    }

    val signature =
        runBlocking {
            key.sign(
                message = Signature.newBuilder().build().enableIdentityText(key = walletPreKey)
            )
        } ?: throw XMTPException("Illegal signature")
    val cipherText = Crypto.encrypt(signature.rawDataWithNormalizedRecovery, bundleBytes)
    return EncryptedPrivateKeyBundle.newBuilder().apply {
        v1 = v1.toBuilder().also { v1Builder ->
            v1Builder.walletPreKey = walletPreKey.toByteString()
            v1Builder.ciphertext = cipherText
        }.build()
    }.build()
}
